### DISCUSSION TITLE
clang-format creation

### DISCUSSION THREAD

**Discussion by Erik-Morbach** [grblHAL/core#192] [General]
Hi,

After forking and modifying the code in some IDEs I realized that it doesn't have a .clang-format file. So every update in the repository generate conflicts based on formating issues.

Creating a .clang-format file and applying it to the project is viable?

**Comment by terjeio** [comment]
> Creating a .clang-format file and applying it to the project is viable?

I do not know. Can be done without changing my coding style?

**Reply by Erik-Morbach** [reply]
I'm trying to come up with a formatting that will not change your code. I will post it here when I finish

**Reply by Erik-Morbach** [reply]
I came up with this configuration after comparing my fork with the official repository and (at least for me) it becomes more readable.
The problem is that the formatting used in the official repository is not global (there are several files where the formatting differs).
Another thing is the usage of #if #else, some of them are indented and others not. some of them have values aligned and others don't. 
As I'm a not pro in clang-formatting, it can be improved a lot. In my perspective, using a .clang-format will make the code a lot more uniform.
[ClangFormat.txt](https://github.com/grblHAL/core/files/9586741/ClangFormat.txt)

**Reply by terjeio** [reply]
> I came up with this configuration after comparing my fork with the official repository and (at least for me) it becomes more readable

Well, that depends on the eyes that see. I find some clang formatted code hard to read and "noisy".

> In my perspective, using a .clang-format will make the code a lot more uniform.

Perhaps so, but it is not a high priority for me. And please do not start submitting PRs with formatting changes only...

I plan to add/modify inline comments in/to [doxygen format](http://svn.io-engineering.com/grblHAL/html/hal_8h.html) - IMO this would be a help for new developers too. I will look at code formatting when I do this, hopefully soon. Your .clang format specification may help with this.

A web frontend for building firmware, at least for the most used drivers, should also be made - compilation likely driven by platformio. It would be nice if someone else than me would do that.

Should I add your .clang format file to the core repo as a suggestion for now?

**Reply by Erik-Morbach** [reply]
> Well, that depends on the eyes that see. I find some clang formatted code hard to read and "noisy".

Yeah, the configuration isn't perfect and there are a lot of issues to adjust, an additional example is the usage of '{' and '}' inside #ifdefs (I don't know if clang can handle all of them properly).

> Should I add your .clang format file to the core repo as a suggestion for now?

If you think it fits the repo and would help, Ok.

> A web frontend for building firmware, at least for the most used drivers, should also be made - compilation likely driven by platformio. It would be nice if someone else than me would do that.

Is there someplace where I can find a kind of specification about this feature? I'm not into web development but could help with something.

**Reply by terjeio** [reply]
> Is there someplace where I can find a kind of specification about this feature?

No, it is just an idea in my head for now. What I am thinking about is to have a list of drivers in a file, likely in the [drivers repo](https://github.com/grblHAL/drivers), with links to files with driver and board properties. Then use these to build a web UI where the user selects what is wanted, from that create a (platformio.ini?) file that is submitted to a server for building the binary and allow the user to download that. It should also be possible for the user to save the options selected. Simple...

The [ESP32 WebUI v3](https://github.com/luc-github/ESP3D-WEBUI/tree/3.0) settings widgets could be used as a basis for this.
